kythe/typescript/indexer.ts

```diff
@@ -216,18 +146,154 @@ class StandardIndexerContext implements IndexerHost {
+
+/**
+ * Determines if a node is a variable-like declaration.
+ */
+function hasExpressionInitializer(node: ts.Node):
+    node is HasExpressionInitializer {
+  return ts.isVariableDeclaration(node) || ts.isParameter(node) ||
+      ts.isBindingElement(node) || ts.isPropertySignature(node) ||
+      ts.isPropertyDeclaration(node) || ts.isPropertyAssignment(node) ||
+      ts.isEnumMember(node);
+}
+
+/**
+ * TSNamespace represents the three declaration namespaces of TypeScript: types,
+ * values, and (confusingly) namespaces. A given symbol may be a type, and/or a
+ * value, and/or a namespace.
+ *
+ * See the table at
+ *   https://www.typescriptlang.org/docs/handbook/declaration-merging.html
+ * for a listing of namespace groups for various declaration types and further
+ * discussion.
+ */
+enum TSNamespace {
+  TYPE,
+  VALUE,
+  NAMESPACE,
+}
+
+/**
+ * Context represents the environment a node is declared in, and may be used for
+ * disambiguating a node's declarations if it has multiple.
+ */
+enum Context {
+  /** No disambiguation about a node's declarations. */
+  Any,
+
+  /** The node is declared as a getter. */
+  Getter,
+
+  /** The node is declared as a setter. */
+  Setter,
+}
+
+/**
+ * A SymbolVNameCache stores a mapping of symbols to the (many) VNames it may
+ * have. Each TypeScript symbol can be be of a different TypeScript namespace
+ * and be declared in a unique context, leading to a total (`TSNamespace` *
+ * `Context`) number of possible VNames for the symbol.
+ *
+ *              TSNamespace    Context
+ *              -----------    -------
+ *              TYPE           Any
+ * ts.Symbol -> VALUE       -> Getter  -> VName
+ *              NAMESPACE      Setter
+ *                             ...
+ *
+ * The `Any` context makes no guarantee of symbol declaration disambiguation.
+ * As a result, unless explicitly set for a given symbol and namespace, the
+ * VName of an `Any` context is lazily set to the VName of an arbitrary context.
+ */
+type NamespaceAndContext = string&{__brand: 'nsctx'};
+class SymbolVNameCache {
+  private readonly cache =
+      new Map<ts.Symbol, Map<NamespaceAndContext, VName>>();
+
+  /**
+   * Serializes a namespace and context as a string to lookup in the cache.
+   *
+   * Each instance of a JavaScript object is unique, so using one as a key fails
+   * because a new object would be generated every time the cache is queried.
+   */
+  private serialize(ns: TSNamespace, context: Context): NamespaceAndContext {
+    return `${ns}${context}` as NamespaceAndContext;
+  }
+
+  /** Get a symbol VName for a given namespace and context, if it exists. */
+  get(symbol: ts.Symbol, ns: TSNamespace, context: Context): VName|undefined {
+    if (this.cache.has(symbol)) {
+      const nsCtx = this.serialize(ns, context);
+      return this.cache.get(symbol)!.get(nsCtx);
+    }
+    return undefined;
+  }
+
+  /** Set a symbol VName for a given namespace and context. */
```

Comment:
This always overrides set names. Is that desired?
I'd expect it to throw if the ts.Symbol and nsCtx were already in the map, as changing it might break other references.

Reply:
Fixed.

Reply:
This could also be added to the docstring for `set`.